### PR TITLE
refactor(rust): use `Result` to return errors instead of `BundleOutput#errors`

### PR DESCRIPTION
**Pull request by hyf0**
### Description

Two behaviors changed:

- Warning won't get emitted if there are errors

~~Not sure if which way is better.~~ Warnings are excluded expectedly. We want to highlight errors in this situation.

- Assets won't be emitted if there are errors

This should be expected.

### CODE REVIEW

**Comment by hyf0** [comment]
* **#2776** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2776?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2776?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2775** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2775?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/rolldown/tests/esbuild/ts/export_type_issue379/artifacts.snap [hunk 1]
```diff
@@ -27,46 +28,3 @@ source: crates/rolldown_testing/src/integration_test.rs
 ───╯
 
 ```
-
```
Comment (hyf0) [inline crates/rolldown/tests/esbuild/ts/export_type_issue379/artifacts.snap:30]: This should also reduce some snapshot noise due to there's no need to inspect bundle output of errored.

**Comment by hyf0** [comment]
### Merge activity

* **Nov 18, 2:38 AM EST**: [Graphite](https://app.graphite.dev/github/pr/rolldown/rolldown/2776) rebased this pull request after merging its parent, because this pull request is set to merge when ready.

**Review by underfin — APPROVED** [review]

File: crates/rolldown/tests/esbuild/importstar_ts/ts_re_export_type_only_file_es6/_config.json [hunk 2]
```diff
@@ -5,7 +5,7 @@
         "name": "entry",
         "import": "entry.ts"
       }
-    ]
-  },
-  "expectError": true
+    ],
+    "shimMissingExports": true
```
Comment (hyf0) [inline crates/rolldown/tests/esbuild/importstar_ts/ts_re_export_type_only_file_es6/_config.json:9]: @IWANABETHATGUY cc. Rolldown doesn't support/detect `export type` natively, using `shimMissingExports` is an workaround for such case and almost have the same behavior as esbuild.

File: crates/rolldown/tests/esbuild/ts/export_type_issue379/artifacts.snap [hunk 3]
```diff
@@ -46,11 +20,15 @@ var b_exports = {};
 __export(b_exports, { foo: () => foo$2 });
 let foo$2 = 123;
 
+//#endregion
+//#region test.ts
+var Test = void 0;
```
Comment (hyf0) [inline crates/rolldown/tests/esbuild/ts/export_type_issue379/artifacts.snap:25]: Negative situation for `shimMissingExports` vs native support for detection on `export type`.

**Comment by hyf0** [comment]
Directly merged due to unstable tests.

**Pull request merged**